libsolidity/analysis/DocStringTagParser.cpp

```diff
@@ -47,6 +47,65 @@ bool DocStringTagParser::parseDocStrings(SourceUnit const& _sourceUnit)
 	return errorWatcher.ok();
 }
 
+bool DocStringTagParser::validateDocStringsUsingTypes(SourceUnit const& _sourceUnit)
+{
+	auto errorWatcher = m_errorReporter.errorWatcher();
+
+	SimpleASTVisitor visitReturns([](ASTNode const&) { return true; }, [&](ASTNode const& _node)
+	{
+		if (auto const* annotation = dynamic_cast<StructurallyDocumentedAnnotation const*>(&_node.annotation()))
+		{
+			auto const& documentationNode = dynamic_cast<StructurallyDocumented const&>(_node);
+
+			size_t returnTagsVisited = 0;
+
+			for (auto const& [tagName, tagValue]: annotation->docTags)
+				if (tagName == "return")
+				{
+					returnTagsVisited++;
+
+					vector<string> returnParameterNames;
+
+					if (auto const* varDecl = dynamic_cast<VariableDeclaration const*>(&_node))
+					{
+						solAssert(varDecl->isPublic(), "@return is only allowed on public state-variables.");
+						returnParameterNames = FunctionType(*varDecl).returnParameterNames();
+					}
+					else if (auto const* function = dynamic_cast<FunctionDefinition const*>(&_node))
+						returnParameterNames = FunctionType(*function).returnParameterNames();
+					else
+						continue;
+
+					string content = tagValue.content;
+					string firstWord = content.substr(0, content.find_first_of(" \t"));
+
+					if (returnTagsVisited > returnParameterNames.size())
+						m_errorReporter.docstringParsingError(
+							2604_error,
+							documentationNode.documentation()->location(),
+							"Documentation tag \"@" + tagName + " " + tagValue.content + "\"" +
+							" exceeds the number of return parameters."
+						);
+					else
+					{
+						string const& parameter = returnParameterNames.at(returnTagsVisited - 1);
+						if (!parameter.empty() && parameter != firstWord)
+							m_errorReporter.docstringParsingError(
+								5856_error,
+								documentationNode.documentation()->location(),
+								"Documentation tag \"@" + tagName + " " + tagValue.content + "\"" +
+								" does not contain the name of its return parameter."
+							);
+					}
+
+				}
+		}
+	});
+
+	_sourceUnit.accept(visitReturns);
+	return errorWatcher.ok();
```

Comment:
If the return value is always `ok()` maybe it would be better not to return anything?

Reply:
`ok()` doesn't always return `true` ;). The method returns the state of the error watcher

Reply:
Oh, good point. It's fine then.